gdnative-sys/build.rs

```diff
@@ -337,26 +337,37 @@ mod api_wrapper {
             let v_min = api.version.minor;
             let gd_api_struct = api.godot_api_struct();
             godot_apis.extend(quote! {
-                let #i = find_api_ptr(core_api_struct, #gd_api_type, #v_maj, #v_min) as *const #gd_api_struct;
+                let #i = find_api_ptr(core_api_struct, #gd_api_type, #v_maj, #v_min)? as *const #gd_api_struct;
             });
             for function in &api.functions {
                 let function_name = function.rust_name();
-                let expect_msg = format!(
+                let message = format!(
                     "API function missing: {}.{}",
                     api.godot_api_struct(),
                     function_name
                 );
                 constructed_struct_fields.extend(quote! {
-                    #function_name: (*#i).#function_name.expect(#expect_msg),
+                    #function_name: Self::map_not_found((*#i).#function_name, #message)?,
                 });
             }
         }
         quote! {
-            pub unsafe fn from_raw(core_api_struct: *const godot_gdnative_core_api_struct) -> Self {
+            fn map_not_found<T>(t: Option<T>, message: &'static str) -> Result<T, InitError> {
+                match t {
+                    Some(t) => Ok(t),
+                    None => Err(InitError::NotFound{message: message.to_string()}),
+                }
+            }
+            unsafe fn from_raw_or_error(core_api_struct: *const godot_gdnative_core_api_struct) -> Result<Self, InitError> {
                 #godot_apis
-                GodotApi{
+                Ok(GodotApi{
                     #constructed_struct_fields
-                }
+                })
+            }
+            pub unsafe fn from_raw(options: *const godot_gdnative_init_options) -> Option<Self> {
+                Self::from_raw_or_error((*options).api_struct)
+                    .map_err(|e| report_error(options, e))
+                    .ok()
             }
```

Comment:
I'm not very sure about the signature for the new `from_raw` method. In the Rust ecosystem, `from_raw` is supposed to perform conversion from a "raw" representation of a certain type (e.g. `Box::from_raw` is a `fn<T>(*mut T) -> Box<T>`), not just any general constructor.

Furthermore, I think the `-sys` crate should not mandate how errors should be handled, if possible. A downstream consumer (including us, in the future) may want to handle the errors in some other way than reporting them immediately: a potentially useful behavior is trying another version of `-sys` (that may support another major version of Godot) before giving up.

Thus, I'd prefer removing the current `from_raw`, and renaming the current `from_raw_or_error` to `from_raw` instead.

Reply:
Because the InitError enum is now public and could be used elsewhere, I changed `InitError::NotFound` to `InitError::Generic`.